File: src/go/cel/cel_test.go
package cel

import (
	"testing"

	"github.com/google/cel-go/cel"
	"github.com/google/cel-go/common/types"
	"github.com/stretchr/testify/assert"
)

func TestCEL(t *testing.T) {
	env, err := cel.NewEnv(
		cel.Variable("account", cel.DynType),
		cel.Variable("transaction", cel.DynType),
	)
	if err != nil {
		t.Fatalf("NewEnv error: %s", err)
	}

	ast, issues := env.Compile(`
		account.balance >= transaction.withdrawal
			|| (account.overdraftProtection
			&& account.overdraftLimit >= transaction.withdrawal - account.balance)
	`)
	if issues != nil && issues.Err() != nil {
		t.Fatalf("Compile error: %s", issues.Err())
	}
	prg, err := env.Program(ast)
	if err != nil {
		t.Fatalf("program construction error: %s", err)
	}

	val, _, err := prg.Eval(map[string]any{
		"account": map[string]any{
			"balance":             500,
			"overdraftProtection": false,
		},
		"transaction": map[string]any{
			"withdrawal": 600,
		},
	})
	if err != nil {
		t.Fatalf("eval error: %s", err)
	}
	assert.Equal(t, val, types.Bool(false))
}

File: src/python/cel/cel_test.py
import unittest

import celpy


class CelTest(unittest.TestCase):
    def test_cel_101(self):
        env = celpy.Environment()
        ast = env.compile(
            """
account.balance >= transaction.withdrawal
    || (account.overdraftProtection
    && account.overdraftLimit >= transaction.withdrawal - account.balance)
"""
        )
        program = env.program(ast)
        result = program.evaluate(
            {
                "account": celpy.json_to_cel(
                    {"balance": 500, "overdraftProtection": False}
                ),
                "transaction": celpy.json_to_cel({"withdrawal": 600}),
            }
        )
        self.assertEqual(result, celpy.celtypes.BoolType(False))


if __name__ == "__main__":
    unittest.main()
